Add session contents for `02: Behind the Scenes` (#5)
unikraft/summer-of-code-2021: content/en/docs/sessions/02-behind-scenes/index.md

---
title: "Session 02: Behind the Scenes"
linkTitle: "02. Behind the Scenes"
---

## Reminders

### Kraft

## Overview

### 01. Virtualization

Through virtualization, multiple operating systems (OS) are able to run on the same hardware, independently, thinking that each one of them controls the entire system.
This can be done using a hypervisor.
There are 2 main virtualized environments: virtual machines and containers.
Unikernels come somewhere between those 2.

#### Virtual Machines

> Virtual machines represent an abstraction of the hardware, over which an operating system can run, thinking that it is alone on the system, and that it controls the hardware below it.

Reviewer: ```suggestion
Virtual machines represent an abstraction of the hardware, over which an operating system can run, thinking that it is alone on the system and that it controls the hardware below it.
```

This is done with the aid of hypervisors, special software components that virtualize the hardware.
There are 2 types of hypervisors, Type 1 and Type 2.
We won't go in depth into them, but it is good to know how they are different:

> * The **Type 1 hypervisor**, also known as as **bare-metal hypervisor**, has direct access to the hardware, and controls all the operating systems that are running on the system.

Reviewer: ```suggestion
* The **Type 1 hypervisor**, also known as as **bare-metal hypervisor**, has direct access to the hardware and controls all the operating systems that are running on the system.
```

This gives it an advantage over classical OSes.
Being a special type of operating system, Unikraft can run bare-metal, or over a hypervisor.

| ![type 1 hypervisor uk](/docs/sessions/02-behind-scenes/images/unikraft1.svg) | ![type 2 hypervisor uk](/docs/sessions/02-behind-scenes/images/unikraft2.svg) |
| :--: | :--: |
| Unikraft over Type 1 hypervisor                                                  | Unikraft over type 2 hypervisor                           		|

### 02. linuxu and KVM

Unikraft can be run in 2 ways:

* As a virtual machine, using QEMU/KVM or Xen.
  It acts as an operating system, having the responsibility to configure the hardware components that it needs (clocks, additional processors, etc).
  This mode gives Unikraft direct and total control over hardware components, allowing advanced functionalities.
* As a `linuxu` build, in which it behaves as a Linux user-space application.
  This severely limits its performance, as everything Unikraft does must go through the Linux kernel, via system calls.
  This mode should be used only for development and debugging.

> When Unikraft is running using QEMU/KVM, it can either be run on an emulated system, or a (para)virtualized one.

Reviewer: ```suggestion
When Unikraft is running using QEMU/KVM, it can either be run on an emulated system or a (para)virtualized one.
```

  In the case of Xen, many of the hardware-related operations must be done through hypercalls, thus reducing the direct interaction of Unikraft with the hardware.
 * [internal libraries](https://github.com/unikraft/unikraft/tree/staging/lib): these define behaviour independent of the hardware, like scheduling, networking, memory allocation, basic file systems.
  These libraries are the same for every platform or architecture, and rely on the platform code and the architecture code to perform the needed actions.
  The internal libraries differ from the external ones in the implemented functionalities.
  The internal ones define parts of the kernel, while the external ones define user-space level functionalities.
  For example, uknetdev and lwip are 2 libraries that define networking components.
  [Uknetdev](https://github.com/unikraft/unikraft/tree/staging/lib/uknetdev) is an internal library that interacts with the network card and defines how packages are sent using it.

>   [Lwip](https://github.com/unikraft/lib-lwip) is an external library, that defines networking protocols, like IP, TCP, UDP.

Reviewer: ```suggestion
  [Lwip](https://github.com/unikraft/lib-lwip) is an external library that defines networking protocols, like IP, TCP, UDP.
```

  It is used for interrupt handling code.
* [newlibc](https://github.com/unikraft/lib-newlib) is the most complete libc currently available for Unikraft, but it still lacks some functionalities, like multithreading.
  newlibc was designed for embedded environments.
 * [musl](https://github.com/unikraft/lib-musl) is, theoretically, the best libc that will be used by Unikraft, but it's currently under testing.

nolibc and isrlib are part of the Unikraft core.
newlibc and musl are external libraries, from the point of view of Unikraft, and they must be included to the build, as shown in Session 01: Baby Steps.

### 05. Configuring Unikraft - Config.uk

Unikraft is a configurable operating system, where each component can be modified, configured, according to the user’s needs.
This Configuration is done using a version of Kconfig, through the **Config.uk** files.
In these files, options are added to enable libraries, applications and different components of the Unikraft core.
The user can then apply those configuration options, using `make menuconfig`, which generates an internal configuration file that can be understood by the build system, **.config**.

> Once configured, the Unikraft image can be built, using make, and run, using the appropriate method (Linux ELF loader, qemu-kvm, xen, others).

Reviewer 2: Insist that there are three ways of configuring an application:
* defining a `Config.uk` file for it with the corresponding options enabled `select CONFIG_FOOBAR`; then run `make menuconfig`, save and that's it (I don't know of an option to do this non-interactively)
* defining a `kraft.yaml` file, adding the configuration options and then running `kraft config`
* using `make menuconfig` and interactively selecting options then saving

Everything results in `.config` being generated, that will be then used by `make` or `kraft build`.

Reviewer 2: And add a tutorial for a simple app (helloworld style) where they configure using all three options. And then throughout the sessions, they do tutorials / challenges using any / all of these options.

Author: I would leave out the first option from tutorials / work items, for now.

### 06. The Bulid System - basics

Once the application is configured, in **.config**, symbols are defined(eg. `CONFIG_ARCH_X86_64`).

> Those symbols are usable both in the C code, to include certain functionalities only if they were selected in the configuring process, and in the actual building process, to include/ exclude source files, or whole libraries.

Reviewer: ```suggestion
Those symbols are usable both in the C code, to include certain functionalities only if they were selected in the configuring process, and in the actual building process, to include / exclude source files, or whole libraries.
```

This last thing is done in **Makefile.uk**, where source code files are added to libraries.
During the build process, all the `Makefile.uk` files (from the Unikraft core and external librarier) are evaluated, and the selected files are compiled and linked, to form the Unikraft image.

| ![unikraft build](/docs/sessions/02-behind-scenes/images/build_uk.svg) |
| :--: 									 |
| The build process of Unikraft 					 |

## Summary

* Unikraft is a special type of operating system, that can be configured to match the needs of a specific application.

> * This configuration is made possible by a system based on Kconfig, that uses **Config.uk** files to add possible configurations, and **.config** files to store the configuration for the build.

Reviewer: ```suggestion
* This configuration is made possible by a system based on Kconfig, that uses **Config.uk** files to add possible configurations, and **.config** files to store the specific configuration for a build.
```

> * The configuration step creates symbols, taht are visible in both Makefiles and source code.

Reviewer: ```suggestion
* The configuration step creates symbols, that are visible in both Makefiles and source code.
```

* Each component has its own **Makefile.uk**, where source files can be added or removed, and be made dependent on the configuration.

> * Unikraft has an internal libc, but it can use other, more complex and complete, libcs, like newlib and musl.

Reviewer: ```suggestion
* Unikraft has an internal libc, but it can use others, more complex and complete,  like newlib and musl.
```

* Being an operating system, it needs to be run by a hypervisor, like KVM, xen, to work at full capacity.
  It can also be run as an ELF, in Linux, but in this way, the true power of Unikraft is not achieved.

## Work Items

### 01. Going through the Code

Go through the [Unikraft core repository](https://github.com/unikraft/unikraft).
Look in the platform code for **linuxu** and **KVM**.

> Spot the differences between what is required for the 2 platforms.

Reviewer 2: As part of the solution folder, give some pointers of what you are supposed to see.

> See what components are initialized in `setup.c`, in both cases.

Reviewer 3: I would give some pointers, such as:

As we can see, the Linuxu version ([plat/linuxu/setup.c](https://github.com/unikraft/unikraft/blob/staging/plat/linuxu/setup.c)) just does a couple of initializations. The kvm version ([plat/kvm/x86/setup.c](https://github.com/unikraft/unikraft/blob/staging/plat/kvm/x86/setup.c)) is more complex, but don't worry, it just does a couple of more initializations. Just take a look at the end of the file, at `_libkvmplat_entry()`, ignore the rest of the file.

--

The kvm file is way too complex for someone such early in the os dev journey to understand imo.

Since this is still an intro into unikraft, we should have a couple of tutorials. The most of participants are probably far away from being independent. Maybe continue with something way easier to understand. Something like:

Shortly, either if we're running on KVM or linuxu, we will reach the same point: booting Unikraft (which is done by calling `ukplat_entry(...);` / `ukplat_entry_argp(...)` in `setup.c` file we've talked about before). Let's take a look now at [lib/ukboot/boot.c](https://github.com/unikraft/unikraft/blob/staging//lib/ukboot/boot.c). This is where these two functions are defined. Let's inspect what `ukplat_entry()` does and focus on just a couple of `if`/`ifdef` guarded blocks:
* In `#if CONFIG_LIBUKALLOC` block, the memory allocator is declared, but it will actually be initialized later in the same function
* in `#if CONFIG_LIBUKSCHED` block, the scheduler is declared, but it will actually be initialized later in the same function
* in `#ifdef CONFIG_LIBUKLIBPARAM` block, we will call a unikraft core library function:  `uk_libparam_parse()`. This will parse command line arguments given to the kernel and pass them further to the libraries.
* in the following couple of (`if` / `elif`) blocks, we try to initialize the allocator declared before. Let's not worry about this for now.
* there will be another `#if CONFIG_LIBUKALLOC` block, but in this one, we initialize the interrupt request (IRQ) subsystem. The IRQ subsystem is necessary to communicate with devices (for example with the network device)
* there will be two `#if CONFIG_LIBUKSCHED` block, but in these ones, the scheduler declared before gets initialized and started.
* in the end, we will just call `main_thread_func(&tma);` which will at some point (not too further away) start our application, but that's a story for another time.

Author: I would rather let them inspect the code themselves and come with questions. I don't want to put too much emphasize on what happens there. They should just find where certain components are initialized, with the aid of the output from ukdebug

### 02. Reminder: Building and Running Unikraft

> Build the core Unikraft image for the **linuxu** and **KVM** platforms, and run them.

Reviewer 2: This is rather unspecific. You need an application to build that.

At the same time, they need some reminders. I would provide them, as a tutorial, **all** steps required to build [app-helloworld](https://github.com/unikraft/app-helloworld) using kraft and make. Make sure there won't be any path related issues in `Makefile` (or they would be using the `UK_...` environment variables). You can copy/adapt from [Session 01: Baby Steps](https://usoc21.unikraft.org/docs/sessions/01-baby-steps/).

> Build the core Unikraft image for the **linuxu** and **KVM** platforms, and run them.

Reviewer: ```suggestion
Build the core Unikraft images for the **linuxu** and **KVM** platforms, and run them.
```

For the linuxu image, it will be like running a Linux ELF. For the KVM image, use `qemu-system`.

> 

Reviewer: ```suggestion
For building, use the manual Makefile based method presented in Session 1.

```

### 03. Different Architectures

> Build the core Unikraft image for the KVM platform, for the **ARM** and **x86** architectures, and run them.

Reviewer 2: They don't have an idea on how to run for **ARM**. I would let them know, as a tutorial. With all instructions and step by step details.

> 

Reviewer: ```suggestion
When running the ARM image using **qemu-system**, you may observe that it gets stuck.
Set a specific `machine` type and `cpu`, using the qemu-system options.
In order to view the available machine types for your version of qemu, run `qemu-system-aarch64 -machine help` (same for cpu, after choosing the machine).
In order to avoid problems, we recommend using the `virt` machine type,
```

Reviewer: I've run into this several times and I think it would be worth it to warn the participants. Also, it is a good place to make them dig a little bit into qemu options and build a command.
However, not all machines and cpus are compatible with running the images ( `-machine virt -cpu cortex-a57` should always work). Alternatively, we can give them the command directly.

### 04. Make It Speak

> Enable the internal debugging library for Unikraft (`ukdebug`), and make it display messages up to debug level.

Reviewer 2: I would once again make this a tutorial. Then, as "tasks":
* tell participants to update the configuration for the log-level
* tell participants to add some debug messages in the `main.c` file

> Enable the internal debugging library for Unikraft (`ukdebug`), and make it display messages up to debug level.

Reviewer: ```suggestion
Enable the internal debugging library for Unikraft (`ukdebug`) and make it display messages up to debug level.
```

Identify which hardware components are initialized for both x86 and ARM, and where.

### 05. Emulation vs Virtualization

> Run the Unikraft core image, for KVM platform, x86 architecture, and measure the runtime, first without using KVM hardware acceleration, then while using it.

Reviewer 2: They don't know how to measure. I would provide them a script:
```
\time -v sudo ...
```
or
```
sudo perf stat ...
```

### 06. Reminder: Using kraft

> Run the [helloworld app](https://github.com/unikraft/app-helloworld), using kraft.

Reviewer 2: This should be moved at the very beginning as a reminder of building (with `kraft` and `make`) from the first session.

> ### 07. Tutorial: Adding Filesystems to an Application

Reviewer 2: Add the entire configuration / resulting files in the `sol/` subfolder (see [session 01](https://github.com/unikraft/summer-of-code-2021/tree/main/content/en/docs/sessions/01-baby-steps/sol)), such that, in the likely case they will run into issues, they will have a reference solution.

> For this tutorial, the aim is to modify the helloworld app, to use the 9pfs filesystem: read something from a file, and display it.

Reviewer 2: I added this phrase. Always start a section by saying what needs to happen as part of it. Start with why, start with the goal, the aim.

> For this tutorial, the aim is to modify the helloworld app, to use the 9pfs filesystem: read something from a file, and display it.

Reviewer: I think the participants may run into a unikraft version issue at this task. Starting from the helloworld application built at task 06, they could try using `~/.unikraft/unikraft` as the core repo for this task too. This will cause linking errors when adding newlib and building the image using `make`, because the unikraft version pulled by kraft is too old (we basically need `staging)`.
A solution to this situation would be to ask them use a fresh clone of the unikraft repo, set on the USoC branch. This may also be required for the 9pfs task in session 1, since using the USoC branch is never mentioned in the session text.

To run Unikraft using qemu-guest, we use:
```
./qemu-guest -e guest_fs/ -k build/app-helloworld_kvm-x86_64
```

If we add the `-D` option, we can see the qemu-system command generated.

You may get the following error:
```
[    0.100664] CRIT: [libvfscore] <rootfs.c @  122> Failed to mount /: 22
```
If you do, check that the mount tag in the configuration is the same as the one used by `qemu-guest`.
`qemu-guest` will use the tag **fs0**.

#### Enter kraft

With kraft, the whole process of configuring, building and running Unikraft can be made easier.
First, we need to modify **kraft.yaml**, to reflect our new configuration, like this:
```
specification: '0.5'
name: helloworld
unikraft:
  version: staging
  kconfig:
    - CONFIG_LIBUK9P=y
    - CONFIG_LIB9PFS=y
    - CONFIG_LIBVFSCORE_AUTOMOUNT_ROOTFS=y
    - CONFIG_LIBVFSCORE_ROOTFS_9PFS=y
    - CONFIG_LIBVFSCORE_ROOTDEV="fs0"
targets:
  - architecture: x86_64
    platform: kvm
libraries:
  newlib:
    version: staging
    kconfig:
      - CONFIG_LIBNEWLIBC=y
volumes:
  fs0:
    driver: 9pfs
    source: guest_fs.tgz
```

> TODO: the archive - I don't know if it's needed to be an archive; it doesn't run either way

Reviewer 2: So this doesn't work? It's unclear from your wording.

Author: No, it doesn't work. Without rebuilding, I can run it with `qemu-guest`, but not with `kraft run`. My guess is that the mount tag used by kraft is not `fs0`, but `kraft run -D` doesn't display anything, so I can't see the command used

Reviewer: I think one way to go is using kraft for building and  `qemu-guest` for running (at least they will learn how to use `kraft.yaml` to configure an application) .

Next, we will make kraft reconfigure our application, using `kraft configure -y`.
In our case, nothing will be modified in **.config**, as we had the same configuration before.
If you get an error like "missing component: newlib", you need to run `kraft list update`.

Then, use
```
kraft build
kraft run
```
You should see(about that...) the contents of you file displayed.

### 08. Give the User a Choice

> Modify **Config.uk**, for the helloworld app, so that the user can choose if the app will display *Hello world*, or what it reads from the file from the previous exercise.

Reviewer 2: Make it a tutorial, it's not trivial.

You will also need to make changes in the source code.

> ### 09. Arguments from Command Line

Reviewer 2: Again, make it a tutorial.

Configure the helloworld app to receive command line arguments and print them.
Send it some arguments.

### 10. Adding a new source file

Create a new source file for your application, and implement a function that sorts a given integer array, by calling qsort, in turn, from different libc variants, and then prints that array.
For each library, check the size of the Unikraft image.

> Enable **nolibc**, and then, as a separate config / build, **newlibc**, both by using **make menuconfig** and modifying **kraft.yaml**.

Reviewer: ```suggestion
Enable **nolibc** and then, as a separate config / build, **newlibc**, both by using **make menuconfig** and modifying **kraft.yaml**.
```

You will have four different configurations and builds:
* nolibc + kraft
* nolibc + make
* newlibc + kraft
* newlibc + make

### 11. More Power to the User

Add the possibility to include the new source file only if a configuration option is selected.

> Make sure than after this change, the application can still be built and run.

Reviewer: ```suggestion
Make sure that after this change, the application can still be built and run.
```

### 12. Less Power to the User

> Delete `Config.uk` and reconfigure / rebuild the app.

Reviewer 2: This is unclear to me. What is the aim? What should they notice?

Author: That the things they added don't happen anymore. I'm not sure, either, about this item.

Reviewer: They could also see that all the applications specific options are removed from the configuration menu.

We do this by running `make menuconfig`, then going to the `Application Options` and enabling our configuration option.

Now we can build and run the new Unikraft image.

### 08. Tutorial: Arguments from Command Line

We want to configure the helloworld app to receive command line arguments and then print them.

For this, the Helloworld application already has a configuration option. 
Run `make menuconfig`, go to `Application Options` and enable `Print arguments`.
If we build and run the image now, using `qemu-guest`, we will see that two arguments are passed to Unikraft: the kernel argument, and a console.
We want to pass it an aditional argument, "foo=bar".

Before this, make sure to reset your configuration, so Unikraft won't use 9pfs for this task, and run `make clean`.

#### Raw qemu command

To send an argument with qemu-system, we use the `-append` option, like this:
```
qemu-system-x86_64 -kernel build/app-helloworld_kvm-x86_64 -append "console=ttyS0 foo=bar" -serial stdio
```

#### qemu-guest script

To send an argument with the qemu-guest script, we use the `-a` option, like this:
```
./qemu-guest -k build/app-helloworld_kvm-x86_64 -a "foo=bar"
```

#### Kraft

> To send an argument while using kraft, we use the `-a` option, like this:

Reviewer: ```suggestion
To send an argument while using kraft:
```
